pytorch_lightning/trainer/training_loop.py

```diff
@@ -504,6 +508,9 @@ def run_training_epoch(self):
                 dataloader_idx,
             )
 
+            # clear memory before running validation
+            del batch
```

Comment:
Well good thing we added it otherwise we wouldn't have noticed.

But I don't understand why it's required now. Does the garbage collector not collect CUDA garbage?

Reply:
it gets marked for garbage collection when the scope of the local variable ends. 
so before, this batch was moved to the device inside the hook call which limited the memory use to that scope. now that I have moved it here, the local variable batch will stay in memory as long as we are in the scope of this training for loop and since we are also running validation here, there is an overlap. 

good news is, in our loop refactor we can again remove the del batch because we will be able to separate the training step from the validation call

Reply:
that being said, GPU parity benchmark is still not happy comparing to PyTorch so I can dig even more this weekend ... :( haha

Reply:
@carmocca Per default it doesn't collect CUDA-garbage. It marks the Tensor object (which is only a pointer to CUDA memory AFAIK) for deletion and PyTorch has it's own garbage collector for CUDA based on this I think.